packages/jsts/src/analysis/projectAnalysis/tsconfigCache.ts

```diff
@@ -0,0 +1,119 @@
+
+    return this.inputFileToTsConfigFilesMap.get(inputFile) ?? null;
+  }
+
+  findTsConfigForInputFile(inputFile: string) {
+    this.pendingTsConfigFiles = this.improvedPendingTsConfigOrder(inputFile);
+
+    debug(`Continuing BFS for file: ${inputFile}, pending order: ${this.pendingTsConfigFiles}`);
+    while (this.pendingTsConfigFiles.length) {
+      const tsConfigPath = this.pendingTsConfigFiles.pop()!;
+      debug(`Computing tsconfig ${tsConfigPath} from bridge`);
+      const tsConfigFile = createProgramOptions(tsConfigPath);
+      tsConfigFile.rootNames.forEach(file => {
+        if (!this.inputFileToTsConfigFilesMap.has(file)) {
+          this.inputFileToTsConfigFilesMap.set(file, tsConfigPath);
+        }
+      });
+      if (tsConfigFile.projectReferences?.length) {
+        info(
+          `Adding referenced project's tsconfigs ${tsConfigFile.projectReferences.map(ref => ref.path)}`,
+        );
+        tsConfigFile.projectReferences
+          .filter(refPath => !this.discoveredTsConfigFiles.has(refPath.path))
+          .forEach(refPath => {
+            this.discoveredTsConfigFiles.add(refPath.path);
+            this.pendingTsConfigFiles.unshift(refPath.path);
+          });
+      }
+      if (this.inputFileToTsConfigFilesMap.has(inputFile)) {
+        const foundTsConfigFile = this.inputFileToTsConfigFilesMap.get(inputFile);
+        info(
+          `Using tsConfig ${foundTsConfigFile} for file source file ${inputFile} (${this.pendingTsConfigFiles.length}/${this.discoveredTsConfigFiles.size} tsconfigs not yet checked)`,
+        );
+        return this.inputFileToTsConfigFilesMap.get(inputFile);
+      }
+    }
+    this.inputFileToTsConfigFilesMap.set(inputFile, null);
+  }
+
+  initializeOriginalTsConfigs(tsconfigs: string[]) {
+    this.initialized = true;
+    this.originalTsConfigFiles = tsconfigs;
+    this.clearFileToTsConfigCache();
+  }
+
+  clearAll() {
+    this.initialized = false;
+    this.originalTsConfigFiles = [];
+    this.clearFileToTsConfigCache();
+  }
+
+  clearFileToTsConfigCache() {
+    this.inputFileToTsConfigFilesMap.clear();
+    this.discoveredTsConfigFiles = new Set<string>(this.originalTsConfigFiles);
+    this.pendingTsConfigFiles = [...this.originalTsConfigFiles];
+  }
+
+  /**
+   * Compute an improved order of the pending tsconfig files with respect to the given inputFile.
+   * This is based on the assumption that a tsconfig *should be* in some parent folder of the inputFile.
+   * As an example, for a file in "/usr/path1/path2/index.js", we would identify look for tsconfig's in the exact
+   * folders * "/", "/usr/", "/usr/path1/", "/usr/path1/path2/" and move them to the front.
+   * Note: This will not change the order between the identified and non-identified tsconfigs.
+   * Time and space complexity: O(n).
+   *
+   * @param inputFile current file to analyze
+   * @return Reordered queue of tsconfig files
+   */
+  private improvedPendingTsConfigOrder(inputFile: string) {
```

Comment:
this should be a verb
```suggestion
  private improvePendingTsConfigOrder(inputFile: string) {
```